Snowflake won't sync without password (user using only private key)

### Describe the bug

If you create a user that is authenticated via private key, the sync will never happen: Metabase will connect to the DW, but the sync will never be able to finish as the connection pool is expecting a password to be created

### To Reproduce

1) create a user following this guide https://docs.snowflake.com/en/user-guide/key-pair-auth
2) connect to Metabase with that users
3) see the error

you'll see that Metabase connects, but the sync never finishes

### Expected behavior

We should use always the same authentication scheme once we connect for all processes

### Logs

_No response_

### Information about your Metabase installation

```JSON
- Metabase version: 1.46.1, but surely comes from 44
```


### Severity

P1

### Additional context

_No response_

Just to add a bit more info, after updating the code to properly retrieve the private key by `:private-key-id` instead of by `:private-key-value` on successive attempts to create a connection spec, we can connect and query using a non-pooled connection, but using the pooled version, e.g. during sync, yields the error:

```
ResourcePool could not acquire a resource from its primary factory or source.
        at com.mchange.v2.resourcepool.BasicResourcePool.awaitAvailable(BasicResourcePool.java:1507)
        at com.mchange.v2.resourcepool.BasicResourcePool.prelimCheckoutResource(BasicResourcePool.java:644)
        at com.mchange.v2.resourcepool.BasicResourcePool.checkoutResource(BasicResourcePool.java:554)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutAndMarkConnectionInUse(C3P0PooledConnectionPool.java:758)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutPooledConnection(C3P0PooledConnectionPool.java:685)
        ... 19 more
Caused by: net.snowflake.client.jdbc.SnowflakeSQLException: JWT token is invalid.
```

I don't think it's a thread safety issue. I am still trying to enable detailed logging from the Snowflake JDBC driver.

--- a/modules/drivers/snowflake/resources/metabase-plugin.yaml
+++ b/modules/drivers/snowflake/resources/metabase-plugin.yaml
@@ -10,7 +10,7 @@ driver:
   connection-properties:
     - name: account
       display-name: Account name
-      helper-text: Enter your Account ID with the region that your Snowflake cluster is running on e.g. "xxxxxxxx.us-east-2.aws". Some regions don't have this suffix.
+      helper-text: Enter your Account ID with the region that your Snowflake cluster is running on e.g. "xxxxxxxx.us-east-2.aws". Some regions don't have this suffix.
       placeholder: xxxxxxxx.us-east-2.aws
       required: true
     - user

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -86,37 +86,27 @@
   Setting the Snowflake driver property privatekey would be easier, but that doesn't work
   because clojure.java.jdbc (properly) converts the property values into strings while the
   Snowflake driver expects a java.security.PrivateKey instance."
-  [{:keys [user password account private-key-value private-key-path private-key-id] :as details}]
-  (cond
-    password
-    details
+  [{:keys [user password account private-key-path]
+    :as   details}]
+  (let [base-details (apply dissoc details (vals (secret/get-sub-props "private-key")))]
+    (cond
+      password
+      details
 
-    private-key-path
-    (let [secret-map       (secret/db-details-prop->secret-map details "private-key")
-          private-key-file (when (some? (:value secret-map))
-                             (secret/value->file! secret-map :snowflake))]
-      (cond-> (apply dissoc details (vals (secret/get-sub-props "private-key")))
-        private-key-file (handle-conn-uri user account private-key-file)))
+      private-key-path
+      (let [secret-map       (secret/db-details-prop->secret-map details "private-key")
+            private-key-file (when (some? (:value secret-map))
+                               (secret/value->file! secret-map :snowflake))]
+        (cond-> base-details
+          private-key-file (handle-conn-uri user account private-key-file)))
 
-    private-key-value
-    (let [private-key-str      (if (bytes? private-key-value)
-                                 (String. ^bytes private-key-value StandardCharsets/UTF_8)
-                                 private-key-value)
-          private-key-file-val (cond-> private-key-str
-                                 (re-find driver.u/data-url-pattern private-key-str) driver.u/decode-uploaded)
-          private-key-file     (secret/value->file! {:connection-property-name "private-key-file"
-                                                     :value                    private-key-file-val})]
-      (handle-conn-uri details user account private-key-file))
-
-    ;; TODO: clean up all of this private key handling logic
-    private-key-id
-    (let [encoded (secret/value->string (secret/latest-for-id private-key-id))
-          decoded (cond-> encoded
-                    (re-find driver.u/data-url-pattern encoded) driver.u/decode-uploaded)
-          file    (secret/value->file! {:connection-property-name "private-key-file"
-                                        :value                    decoded})]
-      (assoc (handle-conn-uri details user account file)
-             :private_key_file file))))
+      :else
+      ;; get-secret-string should get the right value (using private-key-value or private-key-id) and decode it automatically
+      (let [decoded (secret/get-secret-string details "private-key")
+            file    (secret/value->file! {:connection-property-name "private-key-file"
+                                          :value                    decoded})]
+        (assoc (handle-conn-uri base-details user account file)
+               :private_key_file file)))))
 
 (defmethod sql-jdbc.conn/connection-details->spec :snowflake
   [_ {:keys [account additional-options], :as details}]

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -20,7 +20,6 @@
    [metabase.driver.sql.util :as sql.u]
    [metabase.driver.sql.util.unprepare :as unprepare]
    [metabase.driver.sync :as driver.s]
-   [metabase.driver.util :as driver.u]
    [metabase.models.secret :as secret]
    [metabase.query-processor.error-type :as qp.error-type]
    [metabase.query-processor.store :as qp.store]
@@ -34,7 +33,6 @@
    [ring.util.codec :as codec])
   (:import
    (java.io File)
-   (java.nio.charset StandardCharsets)
    (java.sql Connection DatabaseMetaData ResultSet Types)
    (java.time OffsetDateTime ZonedDateTime)))
 

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -83,7 +83,7 @@
   Setting the Snowflake driver property privatekey would be easier, but that doesn't work
   because clojure.java.jdbc (properly) converts the property values into strings while the
   Snowflake driver expects a java.security.PrivateKey instance."
-  [{:keys [user password account private-key-value private-key-path] :as details}]
+  [{:keys [user password account private-key-value private-key-path private-key-id] :as details}]
   (cond
     password
     details
@@ -103,7 +103,17 @@
                                  (re-find driver.u/data-url-pattern private-key-str) driver.u/decode-uploaded)
           private-key-file     (secret/value->file! {:connection-property-name "private-key-file"
                                                      :value                    private-key-file-val})]
-      (handle-conn-uri details user account private-key-file))))
+      (handle-conn-uri details user account private-key-file))
+
+    ;; TODO: clean up all of this private key handling logic
+    private-key-id
+    (let [encoded (secret/value->string (secret/latest-for-id private-key-id))
+          decoded (cond-> encoded
+                    (re-find driver.u/data-url-pattern encoded) driver.u/decode-uploaded)
+          file    (secret/value->file! {:connection-property-name "private-key-file"
+                                        :value                    decoded})]
+      (assoc (handle-conn-uri details user account file)
+             :private_key_file file))))
 
 (defmethod sql-jdbc.conn/connection-details->spec :snowflake
   [_ {:keys [account additional-options], :as details}]

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -76,7 +76,10 @@
                                                           {:user (codec/url-encode user)
                                                            :private_key_file (codec/url-encode (.getCanonicalPath ^File private-key-file))})
         new-conn-uri (sql-jdbc.common/conn-str-with-additional-opts existing-conn-uri :url opts-str)]
-    (assoc details :connection-uri new-conn-uri)))
+    (-> details
+        (assoc :connection-uri new-conn-uri)
+        ;; The Snowflake driver uses the :account property, but we need to drop the region from it first
+        (assoc :account (first (str/split account #"\."))))))
 
 (defn- resolve-private-key
   "Convert the private-key secret properties into a private_key_file property in `details`.

--- a/src/metabase/driver/util.clj
+++ b/src/metabase/driver/util.clj
@@ -427,7 +427,7 @@
                                           ;; version of the -value property (the :type "textFile" one)
                                           (let [textfile-prop (val-kw secrets-server->client)]
                                             (:treat-before-posting textfile-prop)))))
-                         value      (let [^String v (val-kw acc)]
+                         value      (when-let [^String v (val-kw acc)]
                                       (case (get-treat)
                                         "base64" (decode-uploaded v)
                                         v))]

